File: src/09.tree/37.652.ts
import { TreeNode } from "./01.100";

export function findDuplicateSubtrees(
    root: TreeNode | null
): Array<TreeNode | null> {
    const map = new Map<string, number>();
    const ans: Array<TreeNode | null> = [];
    const dfs = (node: TreeNode | null) => {
        if (!node) {
            return " ";
        }

        const val = node.val + "_" + dfs(node.left) + dfs(node.right);

        map.set(val, map.has(val) ? map.get(val) + 1 : 1);

        if (map.get(val) == 2) {
            ans.push(node);
        }

        return val;
    };

    dfs(root);

    return ans;
}

File: src/09.tree/37.652.go
package tree

import "strconv"

func FindDuplicateSubtrees(root *TreeNode) []*TreeNode {
	var dfs func(node *TreeNode) string
	ans := []*TreeNode{}
	m := map[string]int{}

	dfs = func(node *TreeNode) string {
		if node == nil {
			return " "
		}

		key := strconv.Itoa(node.Val) + "_" + dfs(node.Left) + dfs(node.Right)

		m[key]++

		if m[key] == 2 {
			ans = append(ans, node)
		}

		return key
	}

	dfs(root)

	return ans
}
